### ISSUE TITLE
check for empty in Each rule

### ISSUE THREAD

**Issue by batusa** [Respect/Validation#598]
should all of these conditions (we have about 3):

```php
if (empty($input)) {
    return true;
}
```

be removed from the [Each](https://github.com/Respect/Validation/blob/master/library/Rules/Each.php#L37) rule class?

I ran the tests here without these conditions and everything passed, but idk if somehow it could break something. (by removing these conditions the code coverage gets increased a little bit :D)

**Comment by henriquemoody** [comment]
That leads to a bug! Thanks for reporting!

**Comment by batusa** [comment]
while I was creating tests regarding the problem reported in this issue, I ran into other problems. in the docs it says that Each accepts an array or an Iterator, but if I try to run the test below, it fails:

```php
public function testTraversableObjectIsCurrentlyFailing()
{
    $input = new \SplStack();
    $input->push(1);
    $input->push(2);
    $input->push(3);
    $input->push(4);

    $v = new Each(new IntVal());
    $result = $v->validate($input);

    $this->assertTrue($result);
}
```

this is because of the following condition which returns false in case of a Traversable object:
https://github.com/batusa/Validation/blob/fix-each/library/Rules/Each.php#L83
```php
if (!is_array($input) || $input instanceof Traversable) {
    return false;
}
```

this condition is duplicated in `assert`, `check` and `validate`, I''ll fix and extract it to a new method. it would be something like:

```php
protected function isTraversableInput($input)
{
    if (!is_array($input) && !$input instanceof Traversable) {
        return false;
    }

    return true;
}
```

doing so, I'll be able to work only with array or traversable objects.

besides these things, should I `return false` (or throw an exception if the method is assert/check) in case of empty array or traversable object without elements?

**Comment by henriquemoody** [comment]
I think it's the same principles of the `Iterable`, maybe you can use inheritance on that, it would not be wrong.

**Comment by batusa** [comment]
sorry, I didn't get what you meant.

**Comment by henriquemoody** [comment]
The [Each](http://respect.github.io/Validation/docs/each.html) rule needs to iterate over the `$input` to perform the validation, that's why there is the `if (!is_array($input) && !$input instanceof Traversable)`.

What I said is the [Iterable](http://respect.github.io/Validation/docs/iterable.html) performs the exact same validation, for it would be easier for you if you just extend the `Iterable` rule class and use its validation.

**Comment by batusa** [comment]
thanks!! I didn't know about this rule.

living and learning :)

**Comment by batusa** [comment]
I made the changes, everything went fine.

should I use the `count` php function to check for elements?

**Comment by henriquemoody** [comment]
You don't need, just iterate over then. Also, you cannot use `count()` over `Traversable` nor `stdClass()`, so it's unusable for us.

**Comment by batusa** [comment]
you're right. I had forgot that most the Spl* classes implement Countable, that's why `count` was working here with SplStack

**Comment by batusa** [comment]
fixed by #599
